Control Flow Article (#63)
TCCPP/wiki: wiki/cpp-tutorial/control-flow.md

# Control Flow

Control flow is the order in which instructions are executed in a program. You can specify the control flow of a program
by using control flow statements (without these statements the program will execute every instruction one after
another).

| Type of control flow statements | Statements            |
| ------------------------------- | --------------------- |
| Conditional                     | if-else, switch       |
| Loop (Iteration)                | for-loop, while-loop  |
| Jump                            | goto, break, continue |

:::info Block Statement (Compound Statement)

Block `{...}` groups a sequence of statements into a single statement, creating a block scope.

:::

## Conditional Statements

Conditional statements (also known as branching statements or selection statements) allow you to execute different sets
of instructions (branches of code) based on some condition. This condition is any expression that evaluates to a boolean
(`true` or `false`).

### If-Statement

> If-statement executes a substatement if the condition is `true`.

Reviewer: "Substatement" is a fancy word, maybe reword along the lines of
```suggestion
An if statement executes code if the condition is `true`.
```

```cpp
if(condition) {
    /* block executed if condition is true */
}
```

This code updates a `price` of a product where only loyal customers are eligible for a 15% discount (normal customers
pay the regular price):

```cpp
if(eligibleForDiscount) {
    price *= 0.75; /* apply discount of 15% */
}
```

### If-Statement With an Else Branch

> Else statement executes a substatement if condition is `false`. Else statement always goes after the if-statement.

Reviewer: ```suggestion
An else statement executes code if condition is `false`. Else statements always go after an if statement.
```

```cpp
if(condition) {
    /* block executed if condition is true */
} else {
    /* block executed if condition is false */
}
```

Implementation of a program that takes user input and prints if the number is odd or even:

```cpp
#include <iostream>

int main() {
    int number;
    std::cout << "Enter a number: ";
    std::cin >> number;

    /* if the remainder of division by 2 is equal to 0 */
    if(number % 2 == 0) {
        std::cout << "Number is even";
    } else { /* number % 2 == 1 */
        std::cout << "Number is odd";
    }
}
```

:::info Braces are optional

You don't have to use `{...}` (block) when writing if-else statements but it is highly recommended (for better code
readability).

```cpp
if(condition) /* if-statement */;
else /* else-statement */;
```

Example of confusing indentation:

```cpp
if(condition)
    std::cout << "This expression is executed only if the condition is true.";
    std::cout << "This expression is always executed.";
```

:::

### Nested If-Statements

> You can nest if-statements. Most often you will see them nested using the `else if` expression.

Reviewer: ```suggestion
You can nest if statements. Most often you will see them nested in the form of an `else if`:
```

```cpp
if(condition_1) {
    /* 1st block */
    /* condition_1 is true */
} else if(condition_2) {
    /* 2nd block */
    /* condition_1 is false */
    /* condition_2 is true */
} else {
    /* 3rd block */
    /* condition_1 is false */
    /* condition_2 is false */
}
```

this code is equivalent to:

```cpp
if(condition_1) {
    /* 1st block */
} else {
    if(condition_2) {
        /* 2nd block */
    } else {
        /* 3rd block */
    }
}
```

Implementation of a program that prompts user to enter a number, takes user input (number), prints if a number is
positive, negative or a zero:

```cpp
#include <iostream>

int main() {
    int number;
    std::cout << "Enter a number: ";
    std::cin >> number;

    if(number > 0) {
        std::cout << "Positive";
    } else if(number < 0) {
        std::cout << "Negative";
    } else {
        std::cout << "Zero";
    }
}
```

## Loop Statements

The loop statements allow you to execute the same instructions multiple times.

::: info Loop Body

The body of a loop is a common way of referring to a block statement `{...}` that gets executed by the loop.

:::

### While-Loop

> While-loop keeps executing a statement while condition is `true`. Condition gets re-evaluated every iteration.

Reviewer: ```suggestion
A while loop executes its body while its condition is `true`. The condition is checked at the start of every iteration.
```

```cpp
while(condition) {
    /* loop body */
}
```

Implementation of a program printing numbers from 0 to 100 using a while-loop:

```cpp
#include <iostream>

int main() {
    int n = 0;
    while(n <= 100) {
        std::cout << n << '\n';
        ++n;
    }
}
```

Implementation of a program calculating nth [Fibonacci number](https://en.wikipedia.org/wiki/Fibonacci_sequence):
(remember that integers can store only a range of values)

```cpp
#include <iostream>

int main() {
    int a = 0, b = 1, n;
    std::cout << "Enter a number: ";
    std::cin >> n;

    /* F(n + 2) = F(n + 1) + F(n) */
    while(--n) {
        int swap = b;
        b += a;

>         a = swap;

Reviewer: ```suggestion
        int previous_b = b;
        b += a;
        a = previous_b;
```

    }

    std::cout << "The nth Fibonacci number is: " << b;
}
```

### Do-While-Loop

Do-while-loop keeps executing a loop body while condition is `true`. Unlike while-loop it guarantees that the loop body

> gets executed at least once.

Reviewer: ```suggestion
A do-while loop executes its loop body after every iteration, ensuring the loop body runs at least once regardless of any condition.
```

```cpp
do {
    // loop body executed at least once
} while(condition);
```

Implementation of a program that prints out digits of a number:

```cpp
#include <iostream>

int main() {
    int num;
    std::cout << "Enter a number: ";
    std::cin >> num;

    do {
        std::cout << num % 10 << '\n';
        num /= 10;
    } while(num != 0);
}
```

Implementation of a program that sums up user inputted numbers until user inputs `0`:

```cpp
#include <iostream>
int main() {
    int num;
    int sum = 0;
    do {
        std::cin >> num;
        sum += num;
    } while(num != 0);
    std::cout << "The sum is: " << num;
}
```

### For-Loop

For-loop also keeps executing the loop body while condition is `true`. Init statement evaluates exactly once at the
start. Condition also gets re-evaluated every iteration. Expression gets always evaluated after loop body.

```cpp
for(/* init-statement */; /* condition */; /* expression */) {
    /* loop body */
}
```

> For-loop is a nicer way of writing a while-loop and is equivalent to this code:

Reviewer: ```suggestion
A for loop is a nicer way of writing a while loop and is equivalent to this code:
```

```cpp
{
    /* init-statement */
    while (condition) {
        /* statement */
        /* expression */ ;
    }
}
```

This is a valid for-loop that loops indefinitely:

```cpp
for(;;) {
    /* loop body */
}

> ```

Reviewer: Maybe move this towards the end of the section, first let's show what a normal for loop looks like. There should be some explanation about why this is valid.

            std::cout << "Fizz";
        } else if(num % 5 == 0) {
            std::cout << "Buzz";
        } else {
            std::cout << num;
        }
        std::cout << '\n';
    }
}
```

## Jump statements

### Break Statement

The break statement allows you to stop (terminate) the enclosing loop early without checking the condition.
Implementation of a simple guessing game:

```cpp
#include <iostream>

int main() {
    int number = 1234;
    int guess;
    for(;;) { // infinite loop
        std::cout << "Guess a number: ";
        std::cin >> guess;
        if(guess < number) {
            std::cout << "Too low!\n";
        } else if (guess > number) {
            std::cout << "Too high!\n";
        } else {
            std::cout << "You guessed correctly!!!";
            break;
        }
    }
}
```

### Continue Statement

The continue statement allows you to skip the rest of the loop body. The continue statement is often used to reduce
nesting of the statements. Implementation of a program that prints odd numbers:

```cpp
#include <iostream>

int main() {
    for(int i = 0; i < 100; ++i) {
        if(i % 2 == 0) continue;
        std::cout << i << '\n';
    }
}
```

> ### Goto Statement

Reviewer: I'm inclined to say that `goto` should be omitted from this section. Maybe a later article could go over them. They're just very niche - not something a beginner should use or is likely to see.

:::

## Switch Statements

Switch statements are very similar to if-else statements. Condition gets evaluated once and execution jumps to the
appropriate label. Break statement allows you to jump to the end of the switch statement. There are two labels you can
define `case` and `default`. Case label needs to be followed by a constant expression (this restriction allows switch
statements to be optimized by the compiler). If none of the cases match the condition execution jumps to the end of the
switch statement or the default label if one is specified (there can only be one default label).

```cpp
switch (condition) {
case 1:
    std::cout << "One";
    break;
case 2:
    std::cout << "Two";
    [[fallthrough]]; /* Attribute (optional - disables warnings on fallthrough) */
case 3:
    std::cout << "Three";
    break;
default:
    std::cout << "Default"
}
```

Implementation of a minimal calculator that takes in two numbers separated by one of `+-*/` operators.

```cpp
#include <iostream>
#include <print> // C++23 feature

int main() {
    int x, y;
    char op;
    std::cout << "Enter number operator number";
    std::cin >> x >> op >> y;

    switch(op) {

>       case '+':

Reviewer: Make indentation consistent with above
